src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -384,7 +399,53 @@ private void SetVerb()
+
+                    // For most content types and most HTTP methods, we use a callback that lets libcurl
+                    // get data from us if/when it wants it.  However, as an optimization, for POSTs that
+                    // use content already known to be entirely in memory, we hand that data off to libcurl
+                    // ahead of time.  This not only saves on costs associated with all of the async transfer
+                    // between the content and libcurl, it also lets libcurl do larger writes that can, for
+                    // example, enable fewer packets to be sent on the wire.
+                    var inMemContent = _requestMessage.Content as ByteArrayContent;
+                    ArraySegment<byte> contentSegment;
+                    if (inMemContent != null &&
+                        inMemContent.TryGetBuffer(out contentSegment) &&
+                        contentSegment.Count <= InMemoryPostContentLimit) // skip if we'd be forcing libcurl to allocate/copy a large buffer
+                    {
+                        // Only pre-provide the content if the content still has its ContentLength
+                        // and if that length matches the segment.  If it doesn't, something has been overridden,
+                        // and we should rely on reading from the content stream to get the data.
+                        long? contentLength = inMemContent.Headers.ContentLength;
+                        if (contentLength.HasValue && contentLength.GetValueOrDefault() == contentSegment.Count)
+                        {
+                            _inMemoryPostContent = true;
+
+                            // Debug double-check array segment; this should all have been validated by the ByteArrayContent
+                            Debug.Assert(contentSegment.Array != null, "Expected non-null byte content array");
+                            Debug.Assert(contentSegment.Count >= 0, $"Expected non-negative byte content count {contentSegment.Count}");
+                            Debug.Assert(contentSegment.Offset >= 0, $"Expected non-negative byte content offset {contentSegment.Offset}");
+                            Debug.Assert(contentSegment.Array.Length - contentSegment.Offset >= contentSegment.Count,
+                                $"Expected offset {contentSegment.Offset} + count {contentSegment.Count} to be within array length {contentSegment.Array.Length}");
+
+                            // Hand the data off to libcurl with COPYPOSTFIELDS for it to copy out the data. (The alternative
+                            // is to use POSTFIELDS, which would mean we'd need to pin the array in the ByteArrayContent for the
+                            // duration of the request.  Often with a ByteArrayContent, the data will be small and the copy cheap.)
+                            unsafe
+                            {
+                                fixed (byte* inMemContentPtr = contentSegment.Array)
```

Comment:
> Just avoid this code path for empty arrays?

For what benefit?

To avoid this code path for empty arrays, that would mean either a) adding a check on the length and adding code to call SetCurOption with IntPtr.Zero if the length is 0, or b) adding a check on the length as part of the condition for entering this section.  Both of those add a branch.  The former also adds way more IL than is saved and results in the exact same functionality.  The latter saves I think a sum total of 5 IL instructions, basically a null check, at the expense of more C# code, a klunkier access pattern in the fixed statement, and a significantly more expensive code path if the length is 0 (though that should be rare).

This code is insignificant compared to everything else going on around it.  I don't see what this would buy.  What am I missing?